### ISSUE TITLE
support migration conditions api

### ISSUE THREAD

**Issue by cdaringe** [tgriesser/knex#2689]
# Feature discussion / request

1. i want to flag a migration file as `runAlways`, as with other migration softwares (e.g. liquibase)

```ts
// current api, no support
export async function up (knex: Knex) { ... }
```

```ts
// future api, with support
export async function up (knex: Knex) { ... }
up.runAlways = true // or...
up.run = function (): Promise<boolean> { ... }
```

**Comment by kibertoad** [comment]
Can you provide an example use-case when this would be useful?

**Comment by cdaringe** [comment]
yes.  certain `grant all` commands only work against the current state of the database.  instead of duplicating `grant all [...perms to role]` cmds repeatedly in new migration files as new migrations roll in (e.g. new tables created), i want to simple exec it every time at the end of the migration stream.  in this regard, i write the content once.  currently, i have my migration with filename/data `20990101....ts` to force it to always run last, but i don't have a great hook to forcibly rerun that migration every time a migration runs.

iow, duplication `runAlways` from http://www.liquibase.org/documentation/changeset.html

**Comment by elhigu** [comment]
Sounds doable and generally useful feature too. If PR is added for this, with good tests and documentation, I'll be happy to merge it.

**Comment by kibertoad** [comment]
@cdaringe Are you currently actively working on it?

**Comment by cdaringe** [comment]
I am not.  I wish to, but I'm channeling my FOSS cycles to another project ATM

**Comment by tonakai** [comment]
@elhigu hi there it looks like there is not much going on with this one, but it would be useful as well, in the past I've used liquibase which has this feature, and the use case is to run all soft object creations at the of a migration run. They had almost all views as raw and modifying such view in normal ways mean copy/pasting the whole view into a new migration, that makes the source code repo pretty crowded. 
with runalways, views are version controlled like normal source code.
if we can also control the order of scripts running that would be good. i guess we can have one migration file that can run all views as well.

**Comment by kibertoad** [comment]
@tonakai PR is welcome :)

**Comment by mercmobily** [comment]
Thanks for this feature request. The "runAlways" functionality for migrations is indeed useful for maintaining database objects that need to be recreated or updated on every migration run.

**Use Cases**: As you and others have mentioned, this is particularly valuable for:
- Views that need to be recreated when underlying tables change
- Stored procedures and functions that should always reflect the latest version
- Permission grants that need to be reapplied after new objects are created
- Any "soft" database objects that should be version-controlled like regular code rather than treated as one-time migrations

**Current Workarounds**:
1. Using a high-numbered migration file (like `29990101_xxx.js`) to force it to run last - but this doesn't make it actually re-run
2. Manually deleting the migration record from the database migration table
3. Duplicating the same logic in new migration files

**Implementation Considerations**:

The feature would need to support two related but distinct behaviours (similar to Liquibase):
- **runAlways**: Execute on every migration run, regardless of changes
- **runOnChange**: Execute only when the migration content has changed (checksum-based)

For implementation, we could:
1. Add a `runAlways` property to migration objects
2. Skip writing to the migration table for runAlways migrations (as suggested by @elhigu)
3. Or better: track them in the migration table but with a flag indicating they should always run
4. Add ordering control (e.g., `runOrder: 'last'`) to ensure these run after regular migrations

**Comparison with Other Tools**:
- **Liquibase** has both `runAlways` and `runOnChange` attributes, plus `runOrder` for controlling execution sequence
- **Flyway** has "repeatable migrations" (prefixed with `R__`) that run whenever their checksum changes
- **Kysely** doesn't have built-in support but there's a third-party package (`@anchan828/kysely-migration`) that provides repeatable migrations based on checksum comparison

**Proposed API**:

```javascript
// Option 1: Return an object with configuration
exports.up = function(knex) {
  return {
    runAlways: true,
    runOrder: 'last', // optional: 'first' or 'last'
    migration: async () => {
      // Recreate all views
      await knex.raw('CREATE OR REPLACE VIEW user_summary AS ...');
      await knex.raw('GRANT SELECT ON user_summary TO readonly_role');
    }
  };
};

// Option 2: Export a configuration property
exports.config = {
  runAlways: true,
  runOrder: 'last'
};

exports.up = async function(knex) {
  // Standard migration code
};
```

This is a valuable feature that would bring Knex closer to feature parity with other migration tools. A well-tested PR with documentation would be welcome. The implementation should carefully consider:
- How to track these migrations (separate table or flags in existing table)
- Execution order relative to regular migrations
- Clear documentation about use cases and best practices
- Whether to support both runAlways and runOnChange variants

**Related issues:**
- #127 - Extract schema.js functionality for use in migrations
- #4761 - PL/pgSQL basic support (stored procedures)
- #937 - Creating functions and aggregates
